src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp

```diff
@@ -1619,4 +1619,89 @@ namespace hiop
         break;
     } // switch
   } // GramSchmidt
+
+
+
+  hiopLinSolverSymSparseCUSOLVERGPU::hiopLinSolverSymSparseCUSOLVERGPU(const int& n, 
+                                                                       const int& nnz, 
+                                                                       hiopNlpFormulation* nlp)
+    : hiopLinSolverSymSparseCUSOLVER(n, nnz, nlp), 
+      rhs_host_{nullptr},
+      M_host_{nullptr}
+  {
+    rhs_host_ = LinearAlgebraFactory::create_vector("default", n);
+    M_host_ = LinearAlgebraFactory::create_matrix_sparse("default", n, n, nnz);
+  }
+  
+  hiopLinSolverSymSparseCUSOLVERGPU::~hiopLinSolverSymSparseCUSOLVERGPU()
+  {
+    delete rhs_host_;
+    delete M_host_;
+  }
+
+  int hiopLinSolverSymSparseCUSOLVERGPU::matrixChanged()
+  {
+    size_type nnz = M_->numberOfNonzeros();
+    double* mval_dev = M_->M();
+    double* mval_host= M_host_->M();
+
+    index_type* irow_dev = M_->i_row();
+    index_type* irow_host= M_host_->i_row();
+
+    index_type* jcol_dev = M_->j_col();
+    index_type* jcol_host= M_host_->j_col();
+
+    if("device" == nlp_->options->GetString("mem_space")) {
+      checkCudaErrors(cudaMemcpy(mval_host, mval_dev, sizeof(double) * nnz, cudaMemcpyDeviceToHost));
+      checkCudaErrors(cudaMemcpy(irow_host, irow_dev, sizeof(index_type) * nnz, cudaMemcpyDeviceToHost));
+      checkCudaErrors(cudaMemcpy(jcol_host, jcol_dev, sizeof(index_type) * nnz, cudaMemcpyDeviceToHost));      
+    } else {
+      checkCudaErrors(cudaMemcpy(mval_host, mval_dev, sizeof(double) * nnz, cudaMemcpyHostToHost));
+      checkCudaErrors(cudaMemcpy(irow_host, irow_dev, sizeof(index_type) * nnz, cudaMemcpyHostToHost));
+      checkCudaErrors(cudaMemcpy(jcol_host, jcol_dev, sizeof(index_type) * nnz, cudaMemcpyHostToHost));
+    }
+    
+    hiopMatrixSparse* swap_ptr = M_;
+    M_ = M_host_;
+    M_host_ = swap_ptr;
+    
+    int vret = hiopLinSolverSymSparseCUSOLVER::matrixChanged();
+
+    swap_ptr = M_;
+    M_ = M_host_;
+    M_host_ = swap_ptr;
+    
+    return vret;
+  }
+  
+  bool hiopLinSolverSymSparseCUSOLVERGPU::solve(hiopVector& x)
+  {
+    double* mval_dev = x.local_data();
+    double* mval_host= rhs_host_->local_data();
+   
+    if("device" == nlp_->options->GetString("mem_space")) {
+      checkCudaErrors(cudaMemcpy(mval_host, mval_dev, sizeof(double) * n_, cudaMemcpyDeviceToHost));
+    } else {
+      checkCudaErrors(cudaMemcpy(mval_host, mval_dev, sizeof(double) * n_, cudaMemcpyHostToHost));
+    }
+
+    hiopMatrixSparse* swap_ptr = M_;
+    M_ = M_host_;
+    M_host_ = swap_ptr;
```

Comment:
If sparse matrix is implemented correctly, you really should not have to do this. The matrix class needs to provide both, the device data and the host mirror, as well as methods to sync the two.

Reply:
`hiopMatrixSparse` has only got one member function [`M()` to access its data](https://github.com/LLNL/hiop/blob/6b88b876cfa997882f9e812ce61d2ba2132442bd/src/LinAlg/hiopMatrixSparse.hpp#L339), while `hiopMatrixRajaSparseTriplet` has accessors `M()` and `M_host()` for the data on host and device to [see here](https://github.com/LLNL/hiop/blob/6b88b876cfa997882f9e812ce61d2ba2132442bd/src/LinAlg/hiopMatrixRajaSparseTriplet.hpp#L340-L344). 

However, the problem is that previously sparse solver only works on cpu and hybrid. It assumes the KKT matrix `M_` is always on host and has type of `hiopMatrixSparseTriplet`. Therefore in the existing implementation of hiopLinSolverSparseCUSOLVER.cpp, there are many places use host data `M_->M()' and then transfer it to device (for hybrid usage). 
[(for example, see here)](https://github.com/LLNL/hiop/blob/705d8aa193b5af094e4f1d7e7d0925861c3a039d/src/LinAlg/hiopLinSolverSparseCUSOLVER.cpp#L254-L288). The current implementation cannot directly work with matrix on device.

If we want to use the existing `hiopLinSolverSparseCUSOLVER` without adding a new wrapper class, I need to dynamic_cast `M_` to type of `hiopMatrixRajaSparseTriplet` first, sync the data, and change `M_->M()` to `M_->M_host()` in couple of places. This seems more works than adding a new wrapper class.

As a result, we think creating a new wrapper class can minimize the changes to your code, and easier to remove them once the linear solver can support data on device.